File: webapp/go/payment_handler.go
package main

import (
	"net/http"

	"github.com/labstack/echo/v4"
)

type PaymentResult struct {
	TotalTip int64 `json:"total_tip"`
}

func GetPaymentResult(c echo.Context) error {
	ctx := c.Request().Context()

	tx, err := dbConn.BeginTxx(ctx, nil)
	if err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to begin transaction: "+err.Error())
	}
	defer tx.Rollback()

	var totalTip int64
	if err := tx.GetContext(ctx, &totalTip, "SELECT IFNULL(SUM(tip), 0) FROM livecomments"); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to count total tip: "+err.Error())
	}

	if err := tx.Commit(); err != nil {
		return echo.NewHTTPError(http.StatusInternalServerError, "failed to commit: "+err.Error())
	}

	return c.JSON(http.StatusOK, &PaymentResult{
		TotalTip: totalTip,
	})
}

File: webapp/node/src/handlers/payment-handler.ts
import { Context } from 'hono'
import { RowDataPacket } from 'mysql2/promise'
import { HonoEnvironment } from '../types/application'
import { throwErrorWith } from '../utils/throw-error-with'

// GET /api/payment
export const GetPaymentResult = async (
  c: Context<HonoEnvironment, '/api/payment'>,
) => {
  const conn = await c.get('pool').getConnection()
  await conn.beginTransaction()

  try {
    const [[{ 'IFNULL(SUM(tip), 0)': totalTip }]] = await conn
      .query<({ 'IFNULL(SUM(tip), 0)': number } & RowDataPacket)[]>(
        'SELECT IFNULL(SUM(tip), 0) FROM livecomments',
      )
      .catch(throwErrorWith('failed to count total tip'))

    await conn.commit().catch(throwErrorWith('failed to commit'))

    return c.json({ totalTip: totalTip })
  } catch (error) {
    await conn.rollback()
    return c.text(`Internal Server Error\n${error}`, 500)
  } finally {
    await conn.rollback()
    conn.release()
  }
}
